src/modelinstance.cpp

```diff
@@ -593,12 +593,14 @@ Status ModelInstance::waitForLoaded(const uint waitForModelLoadedTimeoutMillisec
 
 void ModelInstance::unloadModel(bool changeState) {
     std::lock_guard<std::recursive_mutex> loadingLock(loadingMutex);
-    this->status.setUnloading();
-    subscriptionManager.notifySubscribers();
-    while (!canUnloadInstance()) {
-        SPDLOG_DEBUG("Waiting to unload model: {} version: {}. Blocked by: {} inferences in progres.",
-            getName(), getVersion(), predictRequestsHandlesCount);
-        std::this_thread::sleep_for(std::chrono::milliseconds(UNLOAD_AVAILABILITY_CHECKING_INTERVAL_MILLISECONDS));
+    if (changeState) {
```

Comment:
As soon as you put unloading state in if this means you are unearthing data races -> below you are resetting stl containers which could be accessed by other PredictRequests/MetadataRequests.

The way to go here (fix for cloud models) would be to:
if (changeState) {
this->status.setUnloading()
} else {
this->status.setLoading()
}
and then go on as usual. This means that we probably should change the name of the variable as well to better reflect what it does.

changeStateToEnd?
changeStateToPermanentEnd?
changeStateToUnwaitable?
isPermanentUnload?